test/functional/mining_timewarp_fork.py

```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test getblocktemplate complies with potential future timewarp-fixing
+   softforks (modulo 600 second nTime decrease).
+
+   We first mine a chain up to the difficulty-adjustment block and set
+   the last block's timestamp 2 hours in the future, then check the times
+   returned by getblocktemplate. Note that we cannot check the actual
+   retarget behavior as difficulty adjustments do not occur on regtest
+   (though it still technically has a difficulty adjustment interval)."""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    connect_nodes_bi,
+)
+
+
+class TimewarpMiningTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 2
+        self.setup_clean_chain = True
+
+    def run_test(self):
+        self.log.info('Create some old blocks')
+        block_1 = self.nodes[0].generate(1)[0]
+        for _ in range(2013):
+            self.nodes[0].generate(1)
+        last_block_time = self.nodes[0].getblockheader(block_1)["time"] + 7199
+
+        self.log.info('Create a block 7199 seconds in the future')
+        self.nodes[0].setmocktime(last_block_time)
+        block_2015 = self.nodes[0].generate(1)[0]
+        assert_equal(self.nodes[0].getblockheader(block_2015)["time"], last_block_time)
+
+        mining_info = self.nodes[0].getmininginfo()
+        assert_equal(mining_info['blocks'], 2015)
+        assert_equal(mining_info['currentblocktx'], 0)
+        assert_equal(mining_info['currentblockweight'], 4000)
+
+        self.restart_node(0)
```

Comment:
Why restart node here? Is this just to reset mocktime? Can you call `setmocktime(0)`?

Reply:
Its just easier than trying to figure out what the current time is and then make sure that its correct?